config/config.go

```diff
@@ -143,17 +143,20 @@ func (c *Config) Ensure(fromCloud bool, logger logging.Logger) error {
 	}
 
 	for idx := 0; idx < len(c.Components); idx++ {
-		dependsOn, err := c.Components[idx].Validate(fmt.Sprintf("%s.%d", "components", idx), resource.APITypeComponentName)
+		component := c.Components[idx]
+		dependsOn, err := component.Validate(fmt.Sprintf("%s.%d", "components", idx), resource.APITypeComponentName)
 		if err != nil {
-			fullErr := errors.Errorf("error validating component %s: %s", c.Components[idx].Name, err)
+			fullErr := errors.Errorf("error validating component %s: %s", component.Name, err)
 			if c.DisablePartialStart {
 				return fullErr
 			}
-			logger.Errorw("component config error; starting robot without component", "name", c.Components[idx].Name, "error", err)
+			logger.Errorw("component config error; starting robot without component", "name", component.Name, "error", err)
+			resLogger := logger.Sublogger(component.ResourceName().String())
```

Comment:
The actual change -- this line and the following `Errorw` call.

Overall, I found the lifetime of the "resource validation and errors we care to output associated with blocks of robot json config" is much larger than the lifetime of the actual resource graph node + logger objects. I consider this duplication of generating a logger a hack, and that we need a better story for validating configs.

Reply:
Ooh I'm confused and probably missing the key point here. Why do you need an identical log on the resource logger _and_ regular logger? From your example in the PR description (and your comments in this PR), the information seems redundant; I thought we'd just replace high-level logs with logs to the `resLogger` where possible.

> "resource validation and errors we care to output associated with blocks of robot json config"

I'm also confused what you mean by the "lifetime" of this and it being "larger" than those of resource graph nodes and their associated loggers. Could you elaborate either here or offline?

Reply:
> I'd be happy to just remove the original log calls and only log to the resource logger.

I'm in favor of removing the original log calls. Even if someone were relying on those specific logs, the log message is still the same from the resource logger. I foresee someone pointing out the duplicative logs as confusing before someone complains about the log now coming from a different logger.

> We haven't created the graph node yet

Ah gotcha... I think we can just continue to use the robot logger when the resource logger is not yet available if that SGTY.

Reply:
>  I think we can just continue to use the robot logger when the resource logger is not yet available if that SGTY.

In my local testing, it was pretty easy to hit these logs lines that prior to graph nodes existing that we definitely want associated with the resource logger namespace.

Thus my comment "the configuration errors we care about is a bigger lifetime than graph nodes".

Reply:
Oh I think I see what you mean now by "the duplication of generating a logger [is] a hack". That "hack" seems fine to me; I cannot imagine that the design of our system will change in the near future such that graph nodes are present and configured for logging in the config validation stage.

Reply:
I'm also in favor of just having the log that is output by `logger.Sublogger`.

As for having to initialize a sub-logger more than once - are there any issues as far as safety (i.e. potentially overwriting logs) or resource usage (i.e. it's expensive to do so) go? If not, I don't think this isn't the worst thing to do.

Reply:
> I'm also in favor of just having the log that is output by logger.Sublogger.

:+1: 

> As for having to initialize a sub-logger more than once - are there any issues as far as safety (i.e. potentially overwriting logs)

No. Two different logger objects with the same "logger namespace" shouldn't have any safety issues. Both will be written (in some order) and the fact that they came from different objects should be indistinguishable.

> or resource usage (i.e. it's expensive to do so) go

Not expensive no. Certainly not expensive when compared to the infrequency of reconfigure and all of the other allocations that does.

Reply:
Cool, then this pattern seems totally fine to me.